include/libassert/platform.hpp

```diff
@@ -0,0 +1,343 @@
+// 202303L then we assume that we are using C++26. Currently, not every compiler has a defined value for C++26
+// and to some degree even checking for 202302L is not 100% reliable as some compilers have chosen to not actually define
+// the version number for C++23 until they are fully compliant. Still this is a rare case and should be fine here.
+// TODO: Once C++26 is fully released we should update this so the else instead uses the proper value.
+#ifndef LIBASSERT_STD_VER
+    #if LIBASSERT_CPLUSPLUS <= 201703L
+        #define LIBASSERT_STD_VER 17
+    #elif LIBASSERT_CPLUSPLUS <= 202002L
+        #define LIBASSERT_STD_VER 20
+    #elif LIBASSERT_CPLUSPLUS <= 202302L
+        #define LIBASSERT_STD_VER 23
+    #else
+        #define LIBASSERT_STD_VER 26 // If our version number is higher than 202303L assume we are using C++26
+    #endif // LIBASSERT_CPLUSPLUS
+#endif // LIBASSERT_STD_VER
+
+
+///
+/// Detect compiler versions.
+///
+
+
+#if defined(__clang__) && !defined(__ibmxl__)
+    #define LIBASSERT_IS_CLANG 1
+    #define LIBASSERT_CLANG_VERSION (__clang_major__ * 100 + __clang_minor__)
+#else
+    #define LIBASSERT_IS_CLANG 0
+    #define LIBASSERT_CLANG_VERSION 0
+#endif
+
+
+#if (defined(__GNUC__) || defined(__GNUG__)) && !defined(__clang__) && !defined(__INTEL_COMPILER)
+    #define LIBASSERT_IS_GCC 1
+    #define LIBASSERT_GCC_VERSION (__GNUC__ * 100 + __GNUC_MINOR__)
+#else
+    #define LIBASSERT_IS_GCC 0
+    #define LIBASSERT_GCC_VERSION 0
+#endif
+
+
+#if defined(_MSC_VER)
+    #define LIBASSERT_IS_MSVC 1
+    #define LIBASSERT_MSVC_VERSION _MSC_VER
+    #include <iso646.h> // alternative operator tokens are standard but msvc requires the include or /permissive- or /Za
+#else
+    #define LIBASSERT_IS_MSVC 0
+    #define LIBASSERT_MSVC_VERSION 0
+#endif
+
+
+///
+/// Detect standard library versions.
+///
+
+// libstdc++
+#ifdef _GLIBCXX_RELEASE
+    #define LIBASSERT_GLIBCXX_RELEASE _GLIBCXX_RELEASE
+#else
+    #define LIBASSERT_GLIBCXX_RELEASE 0
+#endif
+
+
+#ifdef _LIBCPP_VERSION
+    #define LIBASSERT_LIBCPP_VERSION _LIBCPP_VERSION
+#else
+    #define LIBASSERT_LIBCPP_VERSION 0
+#endif
+
+
+#ifdef _MSVC_LANG
+    #define LIBASSERT_CPLUSPLUS _MSVC_LANG
+#else
+    #define LIBASSERT_CPLUSPLUS __cplusplus
+#endif
+
+
+///
+/// Helper macros for compiler attributes.
+///
+
+#ifdef __has_cpp_attribute
+    #define LIBASSERT_HAS_CPP_ATTRIBUTE(x) __has_cpp_attribute(x)
+#else
+    #define LIBASSERT_HAS_CPP_ATTRIBUTE(x) 0
+#endif
+
+
+///
+/// Compiler attribute support.
+///
+
+#if LIBASSERT_HAS_CPP_ATTRIBUTE(nodiscard) && LIBASSERT_STD_VER >= 20
+    #define LIBASSERT_ATTR_NODISCARD_MSG(msg) [[nodiscard(msg)]]
+#else // Assume we have normal C++17 nodiscard support.
+    #define LIBASSERT_ATTR_NODISCARD_MSG(msg) [[nodiscard]]
+#endif
+
+
+#if LIBASSERT_HAS_CPP_ATTRIBUTE(likely)
+    #define LIBASSERT_LIKELY [[likely]]
+#else
+    #define LIBASSERT_ATTR_LIKELY
+#endif
```

Comment:
I'm weary of adding boilerplate that might end up not being used in the project. I think it'd be good to omit these for now, and anything that won't be used in the near future in the library until needed, happy to discuss more.

Reply:
(I comment on `[[likely]]` because it's a micro-optimization that is rarely actually useful, IME. I have similar wariness about things like `LIBASSERT_CPP20_EXPLICIT_BOOL` since while it's cool I'm not sure where in libassert anything would be conditionally explicit (and also its scary to me for explicitness to change between standards))

Reply:
I guess the bigger question is what is less likely? I am not too familiar with the challenges you've encountered through out the library and what is valuable. So some insights into that realm would be appreciated. ^^

Reply:
discussed on discord